src/Client/Connection.cpp

```diff
@@ -573,15 +572,7 @@ void Connection::receiveHello(const Poco::Timespan & handshake_timeout)
     else if (packet_type == Protocol::Server::Exception)
         receiveException()->rethrow();
     else
-    {
-        /// Reset timeout_setter before disconnect,
-        /// because after disconnect socket will be invalid.
-        timeout_setter.reset();
-
-        /// Close connection, to not stay in unsynchronised state.
-        disconnect();
-        throwUnexpectedPacket(packet_type, "Hello or Exception");
-    }
+        throwUnexpectedPacket(timeout_setter, packet_type, "Hello or Exception");
```

Comment:
It doesn't make sense to reset timeout for a connection if the connection is going to be closed, right? So, why bother at all?

Reply:
You are talking about `timeout_setter.reset()` call ?
We need this since disconnect will close fd, while `TimeoutSetter` tries to restore timeout in the dtor

Reply:
Because the `socket` will be invalid, it is reseted here - https://github.com/ClickHouse/ClickHouse/blob/8ed6c034566469360916351665f790fbd6514a0a/src/Client/Connection.cpp#L349-L357

Reply:
Yeh, I see. I'd suggest having additional constructor for TimeoutSetter, which would accept `const std::unique_ptr<Socket> &`, - so it'll be possible to check it. Because passing timeout_setter looks ugly.
But not going to insist on it. Decide yourself.

Reply:
At first, even though I was not sure this looked better, I decided to go with your approach since I did not have a strong opinion on it, and because a) it is at least less hacky, and b) at least one person found it reasonable - so I went ahead with it.

However, after reviewing the code, I realized that it is used not only with `unique_ptr`'s but in only two places overall. Since it is not a frequent use case, I do not think writing additional code is worth it.

That said that I will merge it as-is.